### PR TITLE
KAFKA-7605; Retry async commit failures in integration test cases to fix flaky tests

### PR DESCRIPTION
**Pull request by hachikuji**
We are seeing some timeouts in tests which depend on the `awaitCommitCallback` (e.g. `SaslMultiMechanismConsumerTest.testCoordinatorFailover`). After some investigation, we found that it is caused by a disconnect when attempting the async commit. To fix the problem, we have added simple retry logic to the test utility.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
Well, the one failure in the jdk8 build was the test I'm trying to fix, so just increasing the timeout is not enough 😞. Let me try again to reproduce.

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala [hunk 1]
```diff
@@ -194,8 +194,10 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
   protected def awaitCommitCallback[K, V](consumer: Consumer[K, V],
                                           commitCallback: CountConsumerCommitCallback,
                                           count: Int = 1): Unit = {
-    TestUtils.pollUntilTrue(consumer, () => commitCallback.successCount >= count,
+    val numFailuresBeforePoll = commitCallback.failCount
+    TestUtils.pollUntilTrue(consumer, () => commitCallback.successCount >= count || commitCallback.failCount > numFailuresBeforePoll,
       "Failed to observe commit callback before timeout", waitTimeMs = 10000)
+    assertEquals("Unexpected async commit failure", numFailuresBeforePoll, commitCallback.failCount)
```
Comment (stanislavkozlovski) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: The flakiness comes from the async commit failing?

Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: That's what is timing out. I was wondering if the commit is actually failing.

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
The latest failure confirmed my suspicion:
```
java.lang.AssertionError: Unexpected async commit failure expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at kafka.api.BaseConsumerTest.awaitCommitCallback(BaseConsumerTest.scala:200)
	at kafka.api.BaseConsumerTest.ensureNoRebalance(BaseConsumerTest.scala:216)
	at kafka.api.BaseConsumerTest.testCoordinatorFailover(BaseConsumerTest.scala:117)
```
That likely means that an unexpected rebalance has taken place. I will try to enable some additional logging to get to the cause of the rebalance.

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
At long last, the cause is clear:
```
[2018-11-10 21:29:20,414] ERROR Commit callback failed unexpectedly (kafka.api.SaslMultiMechanismConsumerTest:76)
org.apache.kafka.clients.consumer.RetriableCommitFailedException: Offset commit failed with a retriable exception. You should retry committing the latest consumed offsets.
Caused by: org.apache.kafka.common.errors.DisconnectException
```

I will add some logic to retry when we get `RetriableCommitFailedException`.

**Comment by hachikuji** [comment]
retest this please

**Comment by ijuma** [comment]
So this was just a test bug?

**Comment by hachikuji** [comment]
@ijuma Yep, the test seemed to have regressed in a recent refactor. We need not name the guilty party though 😝

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
Three straight successful jdk8 builds. Before it was failing nearly every time with this particular test case failing every couple times.

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala [hunk 2]
```diff
@@ -191,12 +189,40 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
...
+    }
+
+    class RetryCommitCallback extends OffsetCommitCallback {
+      var isComplete = false
+      var error: Option[Exception] = None
+
+      override def onComplete(offsets: util.Map[TopicPartition, OffsetAndMetadata], exception: Exception): Unit = {
+        exception match {
+          case null =>
+            isComplete = true
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Nit: would it be slightly better if we removed this and in the `case e` clause, changed the last line to `error = Option(e)`?

Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Yes, sounds good.

**Review by ijuma — APPROVED** [review]
Thanks for the PR, LGTM.

**Comment by ijuma** [comment]
Java 11 job failure is unrelated: kafka.server.RequestQuotaTest.testResponseThrottleTimeWhenBothFetchAndRequestQuotasViolated

Java 8 job passed.

**Comment by ijuma** [comment]
@hachikuji does this need to be cherry-picked to 2.1 or any other branch?

**Comment by ijuma** [comment]
I merged to trunk, please cherry-pick if needed @hachikuji.
